service/create/service.go

```diff
@@ -89,45 +96,101 @@ type Service struct {
-			AltNames: []string{
-				"kubernetes",
-				"kubernetes.default",
-				"kubernetes.default.svc",
-				"kubernetes.default.svc.cluster.local",
-			},
-			AllowBareDomains: true,
-			TTL:              "720h",
+		if err := s.createTPR(); err != nil {
+			panic(fmt.Sprintf("could not create cluster resource: %#v", err))
 		}
+		s.Config.Logger.Log("info", "successfully created third-party resource")
+
+		_, certInformer := cache.NewInformer(
+			s.newCertificateListWatch(),
+			&certificatetpr.CustomObject{},
+			resyncPeriod,
+			cache.ResourceEventHandlerFuncs{
+				AddFunc:    s.handleCertAdd,
+				DeleteFunc: s.handleCertDelete,
+				UpdateFunc: s.handleCertUpdate,
+			},
+		)
 
-		// Ensure a PKI backend exists for the cluster.
-		err := s.setupPKIBackend(cert)
-		if err == nil {
-			// Ensure a PKI policy exists for the cluster.
-			err := s.setupPKIPolicy(cert)
-			if err == nil {
-				// PKI setup is OK so attempt to issue a certificate.
-				issueResp, err := s.Issue(cert)
-				if err == nil {
-					s.Config.Logger.Log("info", fmt.Sprintf("cert issued %s %s", cert.CommonName, issueResp.SerialNumber))
-				} else {
-					s.Config.Logger.Log("error", fmt.Sprintf("could not issue cert '%#v'", err))
-				}
-
-			} else {
-				s.Config.Logger.Log("error", fmt.Sprintf("could not setup pki policy '%#v'", err))
-			}
+		s.Config.Logger.Log("info", "starting watch")
 
-		} else {
-			s.Config.Logger.Log("error", fmt.Sprintf("could not setup pki backend '%#v'", err))
-		}
+		// Certificate informer lifecycle can be interrupted by putting a value into a "stop channel".
+		// We aren't currently using that functionality, so we are passing a nil here.
+		certInformer.Run(nil)
 	})
 }
+
+// handleCertAdd issues a certificate using Vault for the certificate TPR. A PKI backend is
+// setup for the Cluster ID if it does not yet exist.
+func (s *Service) handleCertAdd(obj interface{}) {
+	cert := obj.(*certificatetpr.CustomObject)
+	s.Config.Logger.Log("info", fmt.Sprintf("creating certificate '%s'", cert.Spec.CommonName))
+
+	if err := s.setupPKIBackend(cert.Spec); err != nil {
+		s.Config.Logger.Log("error", fmt.Sprintf("could not setup pki backend '%#v'", err))
+		return
+	}
+	if err := s.setupPKIPolicy(cert.Spec); err != nil {
+		s.Config.Logger.Log("error", fmt.Sprintf("could not setup pki backend '%#v'", err))
+		return
+	}
+
+	_, err := s.Issue(cert.Spec)
+	if err != nil {
+		s.Config.Logger.Log("error", fmt.Sprintf("could not issue cert '%#v'", err))
+		return
+	}
+	s.Config.Logger.Log("info", fmt.Sprintf("certificate issued %s", cert.Spec.CommonName))
+}
+
+// handleCertDelete is not yet implemented.
+func (s *Service) handleCertDelete(obj interface{}) {
+	cert := obj.(*certificatetpr.CustomObject)
+	s.Config.Logger.Log("info", fmt.Sprintf("deleting certificate '%s' is not implemented yet", cert.Spec.CommonName))
+}
+
+// handleCertUpdate is not yet implemented.
+func (s *Service) handleCertUpdate(old, cur interface{}) {
```

Comment:
It might make life easier to try and formalise your signatures earlier, so like, I would kind of expect this to return `error` or something, to save you some time in the future. No blocker or anything.